Migrate all remaining TFVM codex into AVM spec (#332)
Azure/Azure-Verified-Modules: docs/content/specs/terraform/_index.md

var.new_network_security_group_name == null ? "${var.subnet_name}-nsg" : var.new_network_security_group_name)
```

<br>

---

<br>

#### ID: TFFR14 - Category: Inputs - No `enabled` or `module_depends_on` variable

> Since Terraform 0.13, `count`, `for_each` and `depends_on` are introduced for modules, module development is significantly simplified. Module's owners **MUST NOT** add variables like `enabled` or `module_depends_on`).

Reviewer: I also am not clear what we mean here - can we provide examples?

Reviewer: ```suggestion
Since Terraform 0.13, `count`, `for_each` and `depends_on` are introduced for modules, module development is significantly simplified. Module's owners **MUST NOT** add variables like `enabled` or `module_depends_on` to control the entire module's operation. Boolean feature toggles are acceptable however.
```

  - `key_vault_network_access` - (Optional) Network access of the key vault Network access of key vault. The possible values are `Public` and `Private`. `Public` means the key vault allows public access from all networks. `Private` means the key vault disables public access and enables private link. Defaults to `Public`.
EOT
}
```

<br>

---

<br>

#### ID: TFNFR18 - Category: Code Style - Every `variable` **MUST** have an appropriate `type`

`type` **MUST** be defined for every `variable`. `type` should be as precise as possible, `any` can only be defined with adequate reasons.

* Use `bool` instead of `string` or `number` for `true/false`
* Use `string` for text
* Use concrete `object` instead of `map(any)`

<br>

---

<br>

#### ID: TFNFR19 - Category: Code Style - `variable` containing confidential data should be declared as `sensitive = true`

If `variable`'s `type` is `object` and contains one or more fields that would be assigned to a `sensitive` argument, then this whole `variable` should be declared as `sensitive = true`, otherwise you should extract sensitive field into separated variable block with `senstive = true`.

<br>

---

<br>

> #### ID: TFNFR20 - Category: Code Style - Declare `nullable = false` when it's possible

Reviewer: Should we state that for simple values, e.g. string, number. Then nullable is ok. Null provides an explicit assertion that the value is not used. Whereas empty string and number values are ambiguous.

However for collection values, e.g. list, set, map. Here nullable should be false and we should use the `[]`, `[]`, `{}` notation respectively?

> #### ID: TFNFR20 - Category: Code Style - Declare `nullable = false` when it's possible

Reviewer: ```suggestion
#### ID: TFNFR20 - Category: Code Style - Declare `nullable = false` when it's possible

Nullable SHOULD be set to `false` for collection values (e.g. sets, maps, lists) when using them in loops. However for scalar values like string and number, a null value MAY have a semantic meaning and as such these values are allowed.
```

<br>

---

<br>

#### ID: TFNFR21 - Category: Code Style - **MUST NOT** declare `nullable = true`

<br>

---

<br>

#### ID: TFNFR22 - Category: Code Style - **MUST NOT** declare `sensitive = false`

<br>

---

<br>

> #### ID: TFNFR23 - Category: Code Style - `variable` with `sensitive = true` **MUST NOT** have default value unless the default value represents turning off a feature, like `default = null` or `default = []`

Reviewer: ```suggestion
#### ID: TFNFR23 - Category: Code Style - `variable` with `sensitive = true` **MUST NOT** have default value unless the default value represents turning off a feature, like `default = null` or `default = []`

Setting a default value for a sensitive input is not permitted, e.g. a default password.
```

> <br>

Reviewer: Have added suggestion above

---

<br>

#### ID: TFNFR24 - Category: Code Style - Deal with deprecated `variable`

Sometimes we will find names for some `variable` are not suitable anymore, or a change should be made to the data type. We want to ensure forward compatibility within a major version, so direct changes are strictly forbidden. The right way to do this is move this `variable` to an independent `deprecated_variables.tf` file, then redefine the new parameter in `variable.tf` and make sure it's compatible everywhere else.

Deprecated `variable` must be annotated as `DEPRECATED` at the beginning of the `description`, at the same time the replacement's name should be declared. E.g.

```terraform
variable "enable_network_security_group" {
  type        = string
  default     = null
  description = "DEPRECATED, use `network_security_group_enabled` instead; Whether to generate a network security group and assign it to the subnet. Changing this forces a new resource to be created."
}
```

A cleanup of `deprecated_variables.tf` can be performed during a major version release.

<br>

---

<br>

> #### ID: TFNFR25 - Category: Code Style - All verified modules **MUST** have `version.tf` file

Reviewer: ```suggestion
#### ID: TFNFR25 - Category: Code Style - All verified modules **MUST** have `terraform.tf` file
```

> `versions.tf` file can only contain one `terraform` block.

Reviewer: Currently we call this file `terraform.tf`.

Would you be prepared to change the name to match the existing standard?

> `versions.tf` file can only contain one `terraform` block.

Reviewer: ```suggestion
`terraform.tf` file can only contain one `terraform` block.
```

The first line of this `terraform` block should define like: `required_version = ">= 1.1"`.

`terraform` block should contain a block called `required_providers`, the content of it is the restrictions for provider's version. Provider's version restriction should be sorted in alphabetical order, same for the assignment statements. `version` restrctions should use `>=` if not specified. No other kinds of restrictions can be used without proper justification.

> All the providers used in the module should be defined in `required_providers`.

Reviewer: ```suggestion
All the providers used in the module should be defined in `required_providers`. ***Please do not add providers that are not directly required by this module. If submodules are used then each submodule should have its own `versions.tf` file.***
```

Good examples:

In verified module:

```terraform
terraform {
  required_providers {
    azurerm = {
      source  = "hashicorp/azurerm"
      version = "~> 3.0"
      configuration_aliases = [ azurerm.alternate ]
    }
  }
}
```

In the root module where we call this verified module:

````terraform
provider "azurerm" {
  features {}
}

provider "azurerm" {
  alias = "alternate"
  features {}
}

module "foo" {
  source = "xxx"
  providers = {
    azurerm = azurerm
    azurerm.alternate = azurerm.alternate
  }
}
````

Bad example:

In verified module:

```terraform
provider "azurerm" {
  # Configuration options
  features {}
}
```

<br>

---

<br>

#### ID: TFNFR28 - Category: Code Style - `output` **MUST** be arranged alphabetically

<br>

---

<br>

#### ID: TFNFR29 - Category: Code Style - `output` contains confidential data should declare `sensitive = true`

<br>

> ---

Reviewer: I think this is clear, what would you like to see?

<br>

#### ID: TFNFR30 - Category: Code Style - Dealing with Deprecated `output`s

> Sometimes we notice that the name of certain `output` is not appropriate anymore, however, since we have to ensure forward compatibility in the same major version, it's not allowed to change the name directly. We need to move it to an independent `deprecated-outputs.tf` file, then redefine a new output in `output.tf` and make sure it's compatible everywhere else in the module.

Reviewer: Can we make this consistent with deprecated variables and use snake case?

Suggest:

```suggestion
Sometimes we notice that the name of certain `output` is not appropriate anymore, however, since we have to ensure forward compatibility in the same major version, it's not allowed to change the name directly. We need to move it to an independent `deprecated_outputs.tf` file, then redefine a new output in `output.tf` and make sure it's compatible everywhere else in the module.
```

> A cleanup can be performed to `deprecated-outputs.tf` and other logics related to compatibility during a major version upgrade.

Reviewer: ```suggestion
A cleanup can be performed to `deprecated_outputs.tf` and other logics related to compatibility during a major version upgrade.
```

<br>

---

<br>

#### ID: TFNFR31 - Category: Code Style - `locals.tf` **MUST** contain only one `locals` block

> All `local` **MUST** be defined in the only `locals` block in `locals.tf` file.

Reviewer: Why is this so prescriptive? Sometimes block comments can be added between distinct locals blocks to add clarity and ease maintenance.

> All `local` **MUST** be defined in the only `locals` block in `locals.tf` file.

Reviewer: ```suggestion
All `local` **MUST** be defined in the `locals.tf` file. Multiple `locals{}` blocks are permitted.
```

Author: I still want to keep `locals.tf` file "pure", it should depends on `variables.tf` only, so we can run some light weight tests by using symbolic links.

I'd like to keep it open to declare `locals` right beneath `resource` and `data` blocks if a `local` referenced these blocks. WDYT?@matt-FFFFFF

Reviewer: I think the use of locals next to resources should be an exception that is only used when we are considering using a fake module to test expressions in `locals.tf` - this is quite an advanced scenario. As long as we add a statement to this effect then I am happy.

<br>

---

<br>

#### ID: TFNFR32 - Category: Code Style - `local` should be arranged alphabetically

<br>

---

<br>

> #### ID: TFNFR33 - Category: Code Style - `local` should use types as precise as possible

Reviewer: We cannot use strict types for locals. I think this needs expanding as the `map(string)` will be confusing for people.

I think this needs context for the use of the local variable, typically from a resource or variable input.

Eg. Type of `object({ name = string, age = number})`:

```terraform
{
  name = "John"
  age  = 52
}
```

> is better than `map(string)`:
> 
> ```terraform
> {
>   name = "John"
>   age  = "52"
> }
> ```

Reviewer: ```suggestion
is better than `object({name = string, age = string})`:

```terraform
{
  name = "John"
  age  = "52"
}
```
```

Author: I've simplified the description.

Reviewer: LGTM
